File: scaly/src/containers/vector.rs
use memory::page::Page;
use std::mem::{align_of, size_of};
use std::ptr::write;

pub struct Vector<T> {
    pub data: *mut T,
    pub length: usize,
}

impl<T> Vector<T> {
    pub fn new(_page: *mut Page, length: usize) -> Vector<T> {
        unsafe {
            Vector {
                data: (*_page).allocate_raw(length * size_of::<T>(), align_of::<T>()) as *mut T,
                length: length,
            }
        }
    }

    pub fn from_raw_array(_page: *mut Page, array: &[T]) -> Vector<T> {
        let vector = Vector::new(_page, array.len());
        let mut address = vector.data;
        for item in array {
            unsafe {
                //write(address, item);
            }
        }
        vector
    }
}

#[test]
fn test_hash_map() {
    use containers::Ref;
    use containers::String;
    use memory::Heap;
    use memory::Page;
    use memory::Region;
    use memory::StackBucket;
    let mut heap = Heap::create();
    let root_stack_bucket = StackBucket::create(&mut heap);
    let root_page = Page::get(root_stack_bucket as usize);
    let r = Region::create_from_page(root_page);
    let _r_1 = Region::create(&r);
    let _vector = Vector::from_raw_array(
        _r_1.page,
        &[
            Ref::new(_r_1.page, String::from_string_slice(_r_1.page, "using")),
            Ref::new(_r_1.page, String::from_string_slice(_r_1.page, "namespace")),
            Ref::new(_r_1.page, String::from_string_slice(_r_1.page, "typedef")),
        ],
    );
}

File: runtime/containers/Vector.cpp
namespace scaly {
namespace containers {

using namespace scaly::memory;

template<class T> struct Array;

template<class T> struct Vector;

template<class T> struct VectorIterator {
    Vector<T>& vector;
    size_t position;

    VectorIterator<T>(Vector<T>& vector) : vector(vector), position(0) {}

    T* next() {
        if (position == vector.length)
            return nullptr;

        position += 1;
        return vector.get(position - 1);
    }
};

template<class T> struct Vector : Object {
    size_t length;
    T* data;

    Vector(T* data, size_t length)
    : length(length), data(data) {}

    Vector<T>(Page* _rp, size_t length) : length(length) {
        if (this->length > 0) {
            this->data = (T*) _rp->allocate_raw(length * sizeof(T), alignof(T));
            std::memset(this->data, 0, length * sizeof(T));
        } else {
            this->data = nullptr;
        }
    }

    Vector<T>(Page* _rp, T* array, size_t length) : length(length) {
        if (this->length > 0) {
            this->data = (T*) _rp->allocate_raw(length * sizeof(T), alignof(T));
            memcpy(this->data, array, length * sizeof(T));
        } else {
            this->data = nullptr;
        }
    }

    Vector<T>(Page* _rp, Array<T>& array) : length(array.length) {
        if (this->length > 0) {
            this->data = (T*) _rp->allocate_raw(length * sizeof(T), alignof(T));
            memcpy(this->data, array.vector->data, length * sizeof(T));
        } else {
            this->data = nullptr;
        }
    }

    T* get(size_t i) {
        if (i >= this->length)
            return nullptr;
        return (*this)[i];
    }

    bool set(size_t i, T item) {
        if (i >= length)
            return false;
        *(*this)[i] = item;
        return true;
    }

    T* operator [](size_t i) {
        if (i >= length)
            return nullptr;

        return data + i;
    }
};

}
}